schematic/store/synapse.py

```diff
@@ -296,6 +296,62 @@ def wrapper(*args, **kwargs):
                     raise ex
         return wrapper
 
+    def send_api_request(self, request_type: str, uri: str, body: Union[Dict, str], headers: Dict = None, endpoint: str = None, **kwargs):
```

Comment:
You might be aware of this, but the synapse client does have these:

```
import synapseclient
syn = synapseclient.login()
syn.restPOST
syn.restGET
syn.restPUT
syn.restDELETE
```

Those are heavily tested, so it might be nice to use those if it captures everything you need.

Reply:
Thanks for mentioning; those methods are what we're interfacing with in this method. Throughout the codebase we've been trying to use general functions where possible to simplify maintenance in the future, and I was trying to continue that here. My thinking was also that when we do the asset store refactor, part of the common store interface will be a method to send an API request to the respective stores. Does this approach make sense or am I overcomplicating things here?

Reply:
Ah - my apologies!  I should have looked at the code closer, but I had mistaken `request` for the `requests` package and I thought you were doing raw `requests.GET/POST/...` calls over the client.

There are multiple ways to think about this.  Since the asset store for 'tables' specifically are things like postgres, synapse tables, Bigquery, naturally it'll be different API calls which might be harder to encapsulate.  In terms of file stores, that would be using something like fsspec (which we do have on our roadmap to build out fs-synapse), which would be pretty different from storing files into tables.

heres some pseudo code for OOP structure.  Some of this could be abstracted in `schematic-db`, so this might be unnecessary.  I encourage everyone to learn about [SOLID principles](https://www.freecodecamp.org/news/solid-principles-explained-in-plain-english/)

```
from abc import ABC
from dataclasses import dataclass

@dataclass
class BaseAssetStore(ABC):
     connection_env: str
     ....
     @abstractmethod
     def store_table(self):
           ....

class SynapseStore(BaseAssetStore):
     def store_table():
           # your implementation here

@dataclass
class BaseTableUpdate(ABC):
    asset_store: BaseAssetStore,   # This is where you put in your store logic for different stores. Those could have different base classes and configurations
    tableToLoad: pd.DataFrame = None, 
    tableName: str = None, 
    datasetId: str = None, 
    existingTableId: str = None,
    restrict: bool = False
    
    @abstract_method
    def createTable():
         ...

@dataclass
class SynapseTableUpdate(BaseTableUpdate)
    asset_store: SynapseStore
    
    def createTable():
         ...
....

```

Reply:
I saw my name mentioned!

I think the eventual asset store refactor will do exactly this @thomasyu888, like schematic db does with [database backends](https://github.com/Sage-Bionetworks/schematic_db/blob/develop/schematic_db/rdb/rdb.py#L39)

Reply:
That's a good point, the calls for other stores will likely be different than what we're using for synapse currently.
@linglp here's a heads up that this will likely change in the future, including a switch to just using the client functions directly at some point

Reply:
Thanks for the example @thomasyu888. I really like the idea of separating asset store base class and table operation. I think it makes sure that each class focuses on one thing. Thanks @GiaJordan for the reminder. I will keep in mind that the structure of the code might be changed in the future.